From 70957583b12819b91f9843849c985514a7023bb5 Mon Sep 17 00:00:00 2001
From: LuK1337 <priv.luk@gmail.com>
Date: Tue, 7 Apr 2020 13:12:18 +0200
Subject: [PATCH] sepolicy: Add sdcard_posix_contextmount_type attribute

* Since we can't use contextmount_type for sdcard_posix
  due to contextmount_type being read only by design we
  need to declare our own attribute to bypass relabelto
  neverallow. That way we can mount external ext4/f2fs
  SD with sdcard_posix context and write permissions.

Test: m -j selinux_policy
Change-Id: I0dfe49cc0b34dfcce2840198843bde1272cbc61c
(cherry picked from commit f8ad2c60111e1218360c8eb05f3eff8991385071)
(cherry picked from commit 4405a1f65f310485365341784c43a1648b69652e)
---
 prebuilts/api/31.0/public/attributes | 3 +++
 prebuilts/api/31.0/public/domain.te  | 2 +-
 2 files changed, 4 insertions(+), 1 deletion(-)

diff --git a/prebuilts/api/31.0/public/attributes b/prebuilts/api/31.0/public/attributes
index b60c9cc62..f4024d7e6 100644
--- a/prebuilts/api/31.0/public/attributes
+++ b/prebuilts/api/31.0/public/attributes
@@ -21,6 +21,9 @@ attribute fs_type;
 # All types used for context= mounts.
 attribute contextmount_type;
 
+# All types used for sdcard_posix context= mounts.
+attribute sdcard_posix_contextmount_type;
+
 # All types used for files that can exist on a labeled fs.
 # Do not use for pseudo file types.
 # On change, update CHECK_FC_ASSERT_ATTRS
diff --git a/prebuilts/api/31.0/public/domain.te b/prebuilts/api/31.0/public/domain.te
index 799a2f1c5..4fa157535 100644
--- a/prebuilts/api/31.0/public/domain.te
+++ b/prebuilts/api/31.0/public/domain.te
@@ -509,7 +509,7 @@ neverallow * rootfs:file { create write setattr relabelto append unlink link ren
 
 # Restrict context mounts to specific types marked with
 # the contextmount_type attribute.
-neverallow * {fs_type -contextmount_type}:filesystem relabelto;
+neverallow * {fs_type -contextmount_type -sdcard_posix_contextmount_type}:filesystem relabelto;
 
 # Ensure that context mount types are not writable, to ensure that
 # the write to /system restriction above is not bypassed via context=
